internal/dataplane/deckgen/consumer_cmp.go

```diff
@@ -0,0 +1,50 @@
+package deckgen
+
+import (
+	"strings"
+
+	"github.com/kong/go-database-reconciler/pkg/file"
+)
+
+type fConsumerByUsernameAndCustomID []file.FConsumer
+
+func (f fConsumerByUsernameAndCustomID) Len() int      { return len(f) }
+func (f fConsumerByUsernameAndCustomID) Swap(i, j int) { f[i], f[j] = f[j], f[i] }
+func (f fConsumerByUsernameAndCustomID) Less(i, j int) bool {
+	if f[i].Username == nil && f[j].Username != nil {
+		return true
+	}
+	if f[i].Username != nil && f[j].Username == nil {
+		return false
+	}
+	if f[i].Username != nil && f[j].Username != nil {
+		switch cmp := strings.Compare(*f[i].Username, *f[j].Username); cmp {
+		case -1:
+			return true
+		case 1:
+			return false
+		case 0:
+			break
+		}
+	}
+
+	// Both usernames are equal, compare custom_id.
+	if f[i].CustomID == nil && f[j].CustomID != nil {
+		return true
+	}
+	if f[i].CustomID != nil && f[j].CustomID == nil {
+		return false
+	}
+	if f[i].CustomID != nil && f[j].CustomID != nil {
+		switch cmp := strings.Compare(*f[i].CustomID, *f[j].CustomID); cmp {
```

Comment:
We do not need to act differently for the 3 cases so I think just use `<=` here is OK.

Reply:
Can you elaborate? In my mind we have to separate those 3 case because:

- if `strings.Compare()` returns -1 we return `true` (`i`th element is "smaller" than `j`th)
- if `strings.Compare()` returns 1 we return `false` (`i`th element is "bigger" than `j`th)
- if `strings.Compare()` returns 0 we continue comparing


We **can** make this shorter into something like:

```
	if f[i].CustomID != nil && f[j].CustomID != nil {
		return strings.Compare(*f[i].CustomID, *f[j].CustomID) == -1
	}
	return false
```

but IMHO that would make this a bit less readable. Additionally if we ever want to extend the comparison criteria it's easy to do: just add these comparisons below vs if we make it shorter then we need to refactor a lot more.

WDYT?

Reply:
Yes, I agree that use the switch-case is clearer that `-1` or `1` stops the comparing, and `0` means we cannot know which one is "less" in the step and may invoke comparing on other attributes.
In this case, I think we could add some comments on the last step.
However, `0` means that we got 2 consumers with the same (non-empty) `username` or `custom_id`, which is invalid. But I believe that it should be processed elsewhere.

Reply:
> However, 0 means that we got 2 consumers with the same (non-empty) username or custom_id, which is invalid. But I believe that it should be processed elsewhere.

Yes, that should be caught somewhere else. Here, we're only focused on sorting the slice of consumers rather than validating it.

Added a comment about that in 39071064eae4dabee4d58ccd428389180e31a7dc